include/macros.h

```diff
@@ -141,6 +141,13 @@
     }                                      \
     (void)0
 
+// horizontal, left to right
+#define PROJECTED_TO_SCREEN_X(projectedPos, invW) (s32)((projectedPos).x * (invW) * (SCREEN_WIDTH / 2) + (SCREEN_WIDTH / 2))
```

Comment:
I don't think there's much value to these macros as it's just mapping the coordinates from the viewport in (-1;1) range to the "screen pixel coords" range (0;width/height)

Also the position argument here isn't projected yet, it's the division by w that does the projection (for example with perspective projection w scales with z)

Reply:
I think you contradicted yourself there: from what I understand, this is a map from projective space to screen coordinates, in the same way as `Actor_ProjectPos` is a map from world coordinates to projective space coordinates? In which case quite a lot of other stuff needs to change name to be clearer: the function, the temps used in the various functions I changed here, etc.

Reply:
What do you call projected coords? I'm refering to it as "coords after projection like usually perspective projection", not "coords in the camera/view frame of reference"

Reply:
Well, based on the name of the argument of `Actor_ProjectPos`, I assumed we were referring to that as projected. And isn't the whole point of w that the camera matrix is for projective space, or what?

Reply:
Coordinates go through:

model space: how they're stored in binary files like objects

world space: after scale, rotation and translation
(these transformation are the "model matrix")

relative to view: after translation/rotation according to camera position (eye, at, up)
("view matrix")

projected: after perspective is applied
("perspective matrix")

---------------

Note that the model and view matrices keep the w of the homogeneous coordinates (x,y,z,w) as 1 all the way through

However at projected, the perspective works by dividing the coordinates by z, this is achieved by making w depend on z.
So using the coordinates as x,y,z or x/w,y/w,z/w makes a big difference

If the result of projecting coordinates (x,y,z,1) is (x',y',z',w'), then x and y are basically the same except that x' and y' range from -1 to 1 ("viewport coordinates" if you forget about w') from one edge of the screen to the other.
The actual projected coordinates on screen are x'/w' and y'/w'.

So you can see x',y' as either projected (if you see them as homogeneous coordinates and don't forget the w') or "viewport" (basically "relative to camera with -1;1 range representing the screen area")
(and idk exactly what z' and z'/w' represent)

------------

In OoT `Play.viewProjectionMtxF` holds the matrix `projectionMtx * viewMtx` and the model matrix just changes constantly, usually using the `Matrix_` system

For `Actor.projectedPos` the game does `(actor->projectedPos, actor->projectedW) = play->viewProjectionMtxF * (actor->world.pos, 1)` 

So my point is in `#define PROJECTED_TO_SCREEN_X(projectedPos, invW) (s32)((projectedPos).x * (invW)`,
`x * invW` is what gives the projected x, and `projectedPos` is not yet projected

So how to name `projectedPos`, I'm not sure. It's weird to refer to it before the divide by w when the intent is to divide by w...

Reply:
I let this stew for a few days and now I think I have an idea what to do,

N64(-era?) terminology seems to conflict with today's, but I didn't look at a whole lot of sources.
http://n64devkit.square7.ch/kantan/step3/1/1_3.htm
https://answers.unity.com/questions/168156/screen-vs-viewport-what-is-the-difference.html

What I (and unity apparently) call:
- [-1;1] from one edge of the screen to the other: viewport
- [0;SCREEN_] from one edge of the screen to the other: screen

What the N64 intro calls these:
- [-1;1] from one edge of the screen to the other: screen
- [0;SCREEN_] from one edge of the screen to the other: device

I haven't seen screen/device anywhere and it's not super consistent with the rest of the manual anyway so let's just roll with viewport/screen

-----------

Back to the PR

instead of
```c
#define PROJECTED_TO_SCREEN_X(projectedPos, invW) (s32)((projectedPos).x * (invW) * (SCREEN_WIDTH / 2) + (SCREEN_WIDTH / 2))
PROJECTED_TO_SCREEN_X(smthPos, invW)
```

I'd suggest

```c
#define VIEWPORT_TO_SCREEN_X(x) (s32)((x) * (SCREEN_WIDTH / 2) + (SCREEN_WIDTH / 2))
VIEWPORT_TO_SCREEN_X(smthPos.x * invW)
```

then the macro doesn't need to care if the input is (perspective-)projected or not, it's just about changing coordinate system

As for what to call `smthPos`, I guess since it goes through the projection matrix let's just call it projectedPos even if it's not used as such (usually not divided by the projectedW) and brings confusion between viewPos and actual projectedPos

Reply:
> N64(-era?) terminology seems to conflict with today's, but I didn't look at a whole lot of sources.

I looked at half a dozen for OpenGL and they were completely inconsistent, so I don't think there is such a thing as standard terminology. As such it's surely better to use the names that the SDK documentation does since that's what people are going to be referring to alongside the codebase.

Reply:
Yeah it's not consistent in the manual either, how surprising

http://n64devkit.square7.ch/kantan/step3/1/1_3.htm as linked above suggests "screen" for [-1;1] and "device" for [0-SCREEN_]
(this is the only reference to "device coordinates" in the whole manual, but this page is referenced in the `gSPViewport` documentation)

http://n64devkit.square7.ch/pro-man/pro14/index.htm (Figure 14-1) and http://n64devkit.square7.ch/n64man/gsp/gSPTextureRectangle.htm "The screen coordinates" suggest "screen" for [0-SCREEN_]
(like the rest of the manual tbh)

http://n64devkit.square7.ch/tutorial/graphics/1/1_6.htm
> The values in this [-1, 1] range are used for the scaling and offset values mentioned previously to convert to screen coordinates. Below is an equation expressing how x coordinate is actually converted into a screen coordinate:
>
> (Equation 1)
> screen_x = x * (SCREEN_WD / 2) + SCREEN_WD / 2

that's enough for me to go with viewport=[-1;1] and screen=[0;SCREEN_]

Reply:
Don't really have too much to say on the actual names, but I do like the idea of the macros just taking a single coordinate, rather than the full `projectedPos` vector.